### PR TITLE
parse headers to struct using reflection tags

### PR DESCRIPTION
**Pull request by bnfinet**
adds `decoder.go`

### CODE REVIEW

**Comment by mozillazg** [comment]
Thanks for your contributor. Could you have time to add some tests for this feature?

File: decode.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	"strconv"
+	"time"
+)
+
+var decoderType = reflect.TypeOf(new(Decoder)).Elem()
+
+// Decoder is an interface implemented by any type that wishes to decode
+// itself from Header fields in a non-standard way.
+type Decoder interface {
+	ParseHeader(header http.Header, v interface{}) error
```
Comment (mozillazg) [inline decode.go]:
```suggestion
	DecodeHeader(header http.Header, v interface{}) error
```

**Comment by bnfinet** [comment]
@mozillazg I've added a test derived from real world usage of Google Calendar's header based notification system.  Thanks for reviewing the PR and considering its inclusion.

File: .bumpversion.cfg [hunk 2]
```diff
@@ -1,7 +1,7 @@
 [bumpversion]
 commit = True
 tag = True
-current_version = 0.2.1
+current_version = 0.3.0
```
Comment (mozillazg) [inline .bumpversion.cfg]:
```suggestion
current_version = 0.2.1
```
The version info will be auto updated via `bumpversion` CLI when bump new release.

File: decode.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	DecodeHeader(header http.Header, v interface{}) error
+}
+
+// DecodeHeader expects to be passed an http.Header and a struct, and parses
+// header into the struct recursively using the same rules as Header (see above)
+func DecodeHeader(header http.Header, v interface{}) error {
+	val := reflect.ValueOf(v)
+	for val.Kind() == reflect.Ptr {
+		val = val.Elem()
+		fmt.Println(val.Kind())
```
Comment (mozillazg) [inline decode.go]:
```suggestion
```

File: encode.go [hunk 4]
```diff
@@ -29,7 +29,7 @@ import (
 const tagName = "header"
 
 // Version ...
-const Version = "0.2.1"
+const Version = "0.3.0"
```
Comment (mozillazg) [inline encode.go]:
```suggestion
const Version = "0.2.1"
```

File: decode_test.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			plrun := *pl
+			plrun.ResourceState = string(tt.args.e)
+			blank := GoogleCalendarPayload{}
+			if err := DecodeHeader(getHeaders(tt.args.e), &blank); (err != nil) != tt.wantErr {
+				t.Errorf("DecodeHeader() error = %v, wantErr %v", err, tt.wantErr)
+			}
```
Comment (mozillazg) [inline decode_test.go:83]:
```suggestion
			}
			if !reflect.DeepEqual(blank, plrun) {
			     t.Errorf("DecodeHeader is not work as expected, got %#v, want %#v", blank, plrun)
		       }
```

File: encode.go (hunk 4 above)
Comment (bnfinet) [reply encode.go]: done

File: decode_test.go (hunk 5 above)
Comment (bnfinet) [reply decode_test.go:83]: done

**Review by mozillazg — APPROVED** [review]
LGTM! Thanks!

**Comment by bnfinet** [comment]
@mozillazg thanks much!

I'm working towards publishing the Google Calendar library.  Do let me know if there's anything that I can do to get `develop` merged to `master`.

**Comment by mozillazg** [comment]
@bnfinet Just want to let you know, the new version of go-httpheader has been released today.

BTW, I have updated some codes you can review these changes via https://github.com/mozillazg/go-httpheader/compare/v0.2.1...v0.3.0 .

**Comment by bnfinet** [comment]
@mozillazg thank you!

That's excellent test coverage.

**Pull request merged**